client/asset/zec/sign.go

```diff
@@ -0,0 +1,256 @@
+//   https://github.com/zcash/zips/blob/main/zip-0243.rst#specification
+//   which extends https://zips.z.cash/zip-0143#specification
+// See also implementation @
+//	https://github.com/iqoption/zecutil/blob/master/sign.go
+func blake2bSignatureHash(
+	subScript []byte,
+	sigHashes *txscript.TxSigHashes,
+	hashType txscript.SigHashType,
+	tx *dexzec.ZecTx,
+	idx int,
+	amt int64,
+	cver [4]byte, // consensus version (little-endian encoded uint32)
+) (preimage, sHash []byte, err error) {
+	if tx.Version < dexzec.VersionSapling {
+		return nil, nil, fmt.Errorf("version %d transactions unsupported", tx.Version)
+	}
+
+	if idx > len(tx.TxIn)-1 {
+		return nil, nil, fmt.Errorf("blake2bSignatureHash error: idx %d but %d txins", idx, len(tx.TxIn))
+	}
+
+	var sigHash bytes.Buffer
+
+	// header is tx.Version with the overwintered flag set.
+	var bVersion [4]byte
+	binary.LittleEndian.PutUint32(bVersion[:], uint32(tx.Version)|(1<<31))
+	sigHash.Write(bVersion[:])
+
+	// nVersionGroupId
+	var groupID uint32 = versionSaplingGroupID
+	if tx.Version == dexzec.VersionCanopy {
+		groupID = versionNU5GroupID
+	}
+	var nVersion [4]byte
+	binary.LittleEndian.PutUint32(nVersion[:], groupID)
+	sigHash.Write(nVersion[:])
+
+	var zeroHash chainhash.Hash
+
+	// hashPrevouts
+	if hashType&txscript.SigHashAnyOneCanPay == 0 {
+		sigHash.Write(sigHashes.HashPrevOuts[:])
+	} else {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// hashSequence
+	if hashType&txscript.SigHashAnyOneCanPay == 0 &&
+		hashType&sigHashMask != txscript.SigHashSingle &&
+		hashType&sigHashMask != txscript.SigHashNone {
+
+		sigHash.Write(sigHashes.HashSequence[:])
+	} else {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// hashOutputs
+	if hashType&sigHashMask != txscript.SigHashSingle && hashType&sigHashMask != txscript.SigHashNone {
+		sigHash.Write(sigHashes.HashOutputs[:])
+	} else if hashType&sigHashMask == txscript.SigHashSingle && idx < len(tx.TxOut) {
+		var (
+			b bytes.Buffer
+			h chainhash.Hash
+		)
+		if err = wire.WriteTxOut(&b, 0, 0, tx.TxOut[idx]); err != nil {
+			return nil, nil, err
+		}
+
+		if h, err = blake2bHash(b.Bytes(), []byte(outputsHashPersonalization)); err != nil {
+			return nil, nil, err
+		}
+		sigHash.Write(h.CloneBytes())
+	} else {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// hashJoinSplits
+	sigHash.Write(zeroHash[:])
+
+	// hashShieldedSpends
+	if tx.Version == dexzec.VersionSapling {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// hashShieldedOutputs
+	if tx.Version == dexzec.VersionSapling {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// nLockTime
+	var lockTime [4]byte
+	binary.LittleEndian.PutUint32(lockTime[:], tx.LockTime)
+	sigHash.Write(lockTime[:])
+
+	// nExpiryHeight
+	var expiryTime [4]byte
+	binary.LittleEndian.PutUint32(expiryTime[:], tx.ExpiryHeight)
+	sigHash.Write(expiryTime[:])
+
+	// valueBalance
+	if tx.Version >= dexzec.VersionSapling {
+		var valueBalance [8]byte
+		binary.LittleEndian.PutUint64(valueBalance[:], 0)
```

Comment:
Peculiar.  Always zero?  Looks like this block would be effectively `sigHash.Write(eightZeros[:])` i.e. removing this `PutUint64` that would seem to be a no-op